pywr-schema/src/nodes/rolling_virtual_storage.rs

```diff
@@ -0,0 +1,247 @@
+/// nodes. This rolling virtual storage node constraints the utilisation of the nodes using a fixed window of the
+/// last `N` days or time-steps. Each time-step the available volume in the virtual storage is based on the maximum
+/// volume less the sum of the utilisation of the nodes over the window. The window is rolled forward each time-step,
+/// with the oldest time-step being removed from the history and the newest utilisation added.
+///
+/// The rolling virtual storage node is useful for representing rolling licences. For example, a 30-day or 90-day
+/// licence on a water abstraction.
+///
+#[derive(serde::Deserialize, serde::Serialize, Clone, Default)]
+pub struct RollingVirtualStorageNode {
+    #[serde(flatten)]
+    pub meta: NodeMeta,
+    pub nodes: Vec<String>,
+    pub factors: Option<Vec<f64>>,
+    pub max_volume: Option<DynamicFloatValue>,
+    pub min_volume: Option<DynamicFloatValue>,
+    pub cost: Option<DynamicFloatValue>,
+    pub initial_volume: Option<f64>,
+    pub initial_volume_pc: Option<f64>,
+    pub window: RollingWindow,
+}
+
+impl RollingVirtualStorageNode {
+    const DEFAULT_ATTRIBUTE: NodeAttribute = NodeAttribute::Volume;
+
+    pub fn add_to_model(
+        &self,
+        network: &mut pywr_core::network::Network,
+        schema: &crate::model::PywrNetwork,
+        domain: &ModelDomain,
+        tables: &LoadedTableCollection,
+        data_path: Option<&Path>,
+        inter_network_transfers: &[PywrMultiNetworkTransfer],
+    ) -> Result<(), SchemaError> {
+        let initial_volume = if let Some(iv) = self.initial_volume {
+            StorageInitialVolume::Absolute(iv)
+        } else if let Some(pc) = self.initial_volume_pc {
+            StorageInitialVolume::Proportional(pc)
+        } else {
+            return Err(SchemaError::MissingInitialVolume(self.meta.name.to_string()));
+        };
+
+        let cost = match &self.cost {
+            Some(v) => v
+                .load(network, schema, domain, tables, data_path, inter_network_transfers)?
+                .into(),
+            None => ConstraintValue::Scalar(0.0),
+        };
+
+        let min_volume = match &self.min_volume {
+            Some(v) => v
+                .load(network, schema, domain, tables, data_path, inter_network_transfers)?
+                .into(),
+            None => ConstraintValue::Scalar(0.0),
+        };
+
+        let max_volume = match &self.max_volume {
+            Some(v) => v
+                .load(network, schema, domain, tables, data_path, inter_network_transfers)?
+                .into(),
+            None => ConstraintValue::None,
+        };
+
+        let node_idxs = self
+            .nodes
+            .iter()
+            .map(|name| network.get_node_index_by_name(name.as_str(), None))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        // The rolling licence never resets
+        let reset = VirtualStorageReset::Never;
```

Comment:
Is there a possibility that a user might want it reset? e.g. a rolling licences that only operates over the summer

Reply:
Though maybe this would be a separate `SeasonalRollingVirtualStorage` scheme type as you'd need away to turn it off

Reply:
I think, for now, this mirrors old behaviour. We can add to this basic functionality in the future.